key-rotator/key/raw.go

```diff
@@ -0,0 +1,233 @@
+var _ encoding.TextMarshaler = Raw{}
+var _ encoding.TextUnmarshaler = &Raw{}
+
+func (r Raw) MarshalBinary() ([]byte, error) {
+	kBytes, err := r.k.MarshalBinary()
+	if err != nil {
+		return nil, fmt.Errorf("couldn't serialize %v key: %w", r.Type(), err)
+	}
+	return append([]byte{byte(r.Type())}, kBytes...), nil
+}
+
+func (r *Raw) UnmarshalBinary(data []byte) error {
+	if len(data) == 0 {
+		return errors.New("empty input")
+	}
+
+	t := Type(data[0])
+	ti := typeInfos[t]
+	if ti == nil {
+		return fmt.Errorf("unknown key type %v (%d)", t, t)
+	}
+	k := ti.newUninitialized()
+	if err := k.UnmarshalBinary(data[1:]); err != nil {
+		return fmt.Errorf("couldn't unmarshal %v key: %w", t, err)
+	}
+	*r = Raw{k}
+	return nil
+}
+
+func (r Raw) MarshalText() ([]byte, error) {
+	binBytes, err := r.MarshalBinary()
+	if err != nil {
+		return nil, err
+	}
+	buf := make([]byte, base64.RawStdEncoding.EncodedLen(len(binBytes)))
+	base64.RawStdEncoding.Encode(buf, binBytes)
+	return buf, nil
+}
+
+func (r *Raw) UnmarshalText(data []byte) error {
+	binBytes := make([]byte, base64.RawStdEncoding.DecodedLen(len(data)))
+	if _, err := base64.RawStdEncoding.Decode(binBytes, data); err != nil {
+		return fmt.Errorf("couldn't decode base64: %v", err)
+	}
+	return r.UnmarshalBinary(binBytes)
+}
+
+func (r Raw) Equal(o Raw) bool {
+	if r.Type() != o.Type() {
+		return false
+	}
+	return r.k.equal(o.k)
+}
+
+// Type returns the type of the raw key.
+func (r Raw) Type() Type { return r.k.keyType() }
+
+// PublicAsCSR returns a PEM-encoding of the ASN.1 DER-encoding of a
+// PKCS#10 (RFC 2986) CSR over the public portion of the key, using the
+// provided FQDN as the common name for the request.
+func (r Raw) PublicAsCSR(csrFQDN string) (string, error) { return r.k.publicAsCSR(csrFQDN) }
+
+// PublicAsPKIX returns a PEM-encoding of the ASN.1 DER-encoding of the
+// public portion of the key in PKIX (RFC 5280) format.
+func (r Raw) PublicAsPKIX() (string, error) { return r.k.publicAsPKIX() }
+
+// AsX962Uncompressed returns a base64 encoding of the X9.62 uncompressed
+// encoding of the public portion of the key, concatenated with the secret
+// "D" scalar.
+func (r Raw) AsX962Uncompressed() (string, error) { return r.k.asX962Uncompressed() }
+
+// AsPKCS8 returns a base64 encoding of the ASN.1 DER-encoding of the key
+// in PKCS#8 (RFC 5208) format.
+func (r Raw) AsPKCS8() (string, error) { return r.k.asPKCS8() }
+
+// raw represents a raw key of one particular type.
+type raw interface {
+	encoding.BinaryMarshaler
+	encoding.BinaryUnmarshaler
+
+	// keyType returns the type of the raw key.
+	keyType() Type
+
+	// equal determines if this raw is equal to the given raw, which can be
+	// assumed to be of the same key type.
+	equal(o raw) bool
+
+	// publicAsCSR returns a PEM-encoding of the ASN.1 DER-encoding of a
+	// PKCS#10 (RFC 2986) CSR over the public portion of the key, using the
+	// provided FQDN as the common name for the request.
```

Comment:
nit: mention that the CSR will be signed with the private portion of the key (right?)

Reply:
Yep, it's signed (at least per the documentation of `x509.CreateCertificateRequest`). Updated comments.